### PR TITLE
Allow passing an array to library(vs_module_defs)

### PR DESCRIPTION
**Pull request by dcbaker**
There are some cases where we want to use a def file with one compiler,
but not another. For example, you might want to use one with msvc,
clang-cl, and intel-cl, but not with mingw. Previously this required
duplicating the library call, one for the compilers with the def file,
and one without. This allows us to pass an array, possibly of zero
elements.

### CODE REVIEW

**Comment by jon-turney** [comment]
Did you consider allowing an empty string to have this special meaning?

**Comment by dcbaker** [comment]
I thought about it, but in most cases we use an empty array to mean "no value"

**Comment by jon-turney** [comment]
> but in most cases we use an empty array to mean "no value"

True (except where we don't) :)

I guess what I find irrregular about this is the 0 or 1-item list part.  How about permitting `vs_module_defs:` to be (something that can be coerced to) a file, `[]`, or maybe `''`?

**Comment by dcbaker** [comment]
Here's my thought on the 1 or 0 length thing, this should work:
```meson
def = []
# our .def file isn't used with mingw
if host_machine.system() == 'windows' and cc.get_id() != 'gcc'
    def += 'def.def'
endif
```
Otherwise you'll get an annoying error about how `[]` is valid, and `def.def` is valid, but `['def.def']` is not.

**Comment by jon-turney** [comment]
> this should work

Well, perhaps.  I think it would be more normal to write `=` rather than `+=` ,  since it doesn't make much sense to construct `vs_module_defs:` by concatenation, since it can only be a single value?

I'm just worried that this is a bit inconsistent with other places where values and empty lists are accepted, but non-empty lists aren't (e.g `library(name_suffix:)`?

Or if list flattening everywhere means that values and single-item lists are syntactically equivalent, then that should be mentioned somewhere other than the documentation for this keyword.

**Comment by nirbheek** [comment]
The other day I was thinking how we should standardize on a value that signifies "use the default value for this kwarg", similar to how in Python you can pass `None` as the value for all kwargs to mark them as 'unset'.

`[]` is not very discoverable for that (and I say this as the person who started this trend in the first place...). `false` is slightly better, but it is also a valid value for a kwarg. Maybe we want to make a new constant?

Another question is, why not use something like

```meson
library_kwargs = {}
if cc.get_id() == 'msvc'
  library_kwargs += {'vs_module_defs': 'libfoo.def'}
endif
library(... kwargs: library_kwargs)
```

This does the same thing without duplicating the `library()` call.

**Comment by jpakkane** [comment]
Much like `disabler` we could perhaps have a `kwargunsetter` object, whose only use is that if it is passed to a kwarg, the outcome is the same as if the kwarg had not been written in the code at all?

**Comment by nirbheek** [comment]
> Much like `disabler` we could perhaps have a `kwargunsetter` object, whose only use is that if it is passed to a kwarg, the outcome is the same as if the kwarg had not been written in the code at all?

We could generalize it to have an object `unsetter()` which unsets whatever you set it to, whether that's a variable or a kwarg.

**Comment by jpakkane** [comment]
That would be problematic in the following case:

```
opt_kwarg = something
if something_else
   opt_kwarg = unsetter()
endif
some_function(..., foo: opt_kwarg)
```

If unsetting unsets the variable, then the last line leads to a missing variable error.

**Comment by nirbheek** [comment]
True, that's a bad interaction. I'm not sure I like `kwargunsetter()` as-is. Feels like a super-specific thing that already has a different solution.

**Comment by jon-turney** [comment]
> If unsetting unsets the variable, then the last line leads to a missing variable error.

Not intractable, if

```
opt_kwarg = something
if something_else
   opt_kwarg = none()
endif
some_function(..., foo: opt_kwarg)
```
where `none()` returns a value which is distinguished from all other values, and a kwarg with that value is defined to behave the same as if was absent.

But I'm not sure this is an improvement.  How is `[none()]` or `['something', none()]` supposed to behave? and this has drifted pretty far from the original problem, which already has a solution with kwarg dicts.

**Comment by xclaesse** [comment]
> ```
> opt_kwarg = something
> if something_else
>    opt_kwarg = none()
> endif
> some_function(..., foo: opt_kwarg)
> ```

Can be done, but not pretty:
```meson
d = {'foo': something}
if something_else
  d = {}
endif
some_function(..., kwargs: d)
```

**Comment by bonzini** [comment]
FWIW I had a similar issue with `static: false` being different from nothing at all, and solved it with `kwargs`.

**Comment by dcbaker** [comment]
I think we'd discussed having a `none()` or `default()` for solving some of these cases. I find the `kwargs` thing really hard to follow as it adds another level of indirection, but this clearly is not the solution we want.
